# Electricity Demand Forecasting Dataset

## Data Sources

The dataset is collected from multiple reliable sources:

-  **Power Grid Company of Bangladesh (PGCB)**  
  Provides:
  - Electricity generation data  
  - Demand information  
  - Load shedding records  
  - Fuel mix (gas, coal, imports, etc.)

-  **Weather Data Sources**  
  Provides:
  - Temperature  
  - Humidity  
  - Rainfall  
  - Wind speed  

---

##  Data Integration

All datasets are merged using the `datetime column`.

This allows us to:
- Align power and weather data accurately  
- Perform time-based analysis  
- Understand real-world patterns in electricity usage  

---

##  Objectives

### 1.  Forecast Electricity Demand
- Predict future `demand_mw` values  
- Help in planning electricity supply  

### 2.  Estimate Actual Demand
- Include `load shedding` to estimate real demand  
- Understand hidden or unmet demand  

### 3.  Analyze Key Impacts

 
####  Generation Mix
- Contribution of:
  - Gas  
  - Coal  
  - Imports  
  - Other sources  

---

##  Purpose

- Support **energy planning**  
- Improve **grid stability**  
- Enable **data-driven decision making** in the power sector


In [109]:
# import dependencies
import pandas as pd
import numpy as np

# Dataset Loading

In [110]:
# dataset load
df = pd.read_csv("electricity_demand_weather_merged.csv")

In [111]:
df.head()

,datetime,generation_mw,demand_mw,load_shedding,gas,liquid_fuel,coal,hydro,solar,wind,india_bheramara_hvdc,india_tripura,india_adani,nepal,remarks,temperature,humidity,rain,windspeed
0,2020-01-01 00:00:00,5913.0,5913,0,5493,0,160,47,0.0,NaN,129,76,NaN,NaN,NaN,18.3,74.0,0.0,3.5
1,2020-01-01 01:00:00,5570.0,5570,0,5117,0,160,47,0.0,NaN,129,74,NaN,NaN,NaN,16.5,81.0,0.0,6.0
2,2020-01-01 02:00:00,5351.0,5351,0,4900,0,159,47,0.0,NaN,129,72,NaN,NaN,NaN,14.6,89.0,0.0,8.4
3,2020-01-01 03:00:00,5223.0,5223,0,4773,0,160,46,0.0,NaN,129,70,NaN,NaN,NaN,14.5,90.0,0.0,6.3
4,2020-01-01 04:00:00,5142.0,5142,0,4697,0,159,46,0.0,NaN,129,68,NaN,NaN,NaN,14.1,90.0,0.0,8.2


In [112]:
df.tail()

,datetime,generation_mw,demand_mw,load_shedding,gas,liquid_fuel,coal,hydro,solar,wind,india_bheramara_hvdc,india_tripura,india_adani,nepal,remarks,temperature,humidity,rain,windspeed
49847,2025-06-16 20:00:00,15000.0,15035,33,6361,1985,4168,172,0.0,64.0,924,26,1262.0,38.0,NaN,27.6,89.0,0.0,14.3
49848,2025-06-16 21:00:00,14971.0,15000,28,6373,1961,4134,172,0.0,59.0,924,40,1270.0,38.0,NaN,27.2,89.0,0.0,15.2
49849,2025-06-16 22:00:00,15109.0,15130,20,6421,1955,4228,172,0.0,45.0,924,36,1290.0,38.0,NaN,27.0,90.0,0.0,14.2
49850,2025-06-16 23:00:00,15165.0,15205,38,6437,1984,4268,172,0.0,41.0,924,42,1259.0,38.0,NaN,26.8,91.0,0.0,13.7
49851,2025-06-17 00:00:00,14628.0,14650,21,6302,1672,4186,172,0.0,41.0,924,42,1251.0,38.0,NaN,26.7,91.0,0.1,14.3


In [113]:
df.columns

Index(['datetime', 'generation_mw', 'demand_mw', 'load_shedding', 'gas',
       'liquid_fuel', 'coal', 'hydro', 'solar', 'wind', 'india_bheramara_hvdc',
       'india_tripura', 'india_adani', 'nepal', 'remarks', 'temperature',
       'humidity', 'rain', 'windspeed'],
      dtype='object')

In [114]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49852 entries, 0 to 49851
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   datetime              49852 non-null  object 
 1   generation_mw         49852 non-null  float64
 2   demand_mw             49852 non-null  int64  
 3   load_shedding         49852 non-null  int64  
 4   gas                   49852 non-null  int64  
 5   liquid_fuel           49852 non-null  int64  
 6   coal                  49852 non-null  int64  
 7   hydro                 49852 non-null  int64  
 8   solar                 49852 non-null  float64
 9   wind                  18664 non-null  float64
 10  india_bheramara_hvdc  49852 non-null  int64  
 11  india_tripura         49852 non-null  int64  
 12  india_adani           7326 non-null   float64
 13  nepal                 5339 non-null   float64
 14  remarks               3794 non-null   object 
 15  temperature        

In [115]:
df.shape

(49852, 19)

# Data Description Table

| Column Name              | Data Type | Description |
|-------------------------|----------|-------------|
| datetime                | object   | Timestamp (needs conversion to datetime) |
| generation_mw           | float64  | Total electricity generation |
| demand_mw               | int64    | supplied (served demand) |
| load_shedding           | int64    | Unserved demand (power shortage) |
| gas                     | int64    | Gas-based generation |
| liquid_fuel             | int64    | Furnace oil / diesel generation |
| coal                    | int64    | Coal-based generation |
| hydro                   | int64    | Hydropower generation |
| solar                   | float64  | Solar generation |
| wind                    | float64  | Wind generation (many missing values) |
| india_bheramara_hvdc    | int64    | Power import via HVDC |
| india_tripura           | int64    | Import from Tripura |
| india_adani             | float64  | Import from Adani (sparse) |
| nepal                   | float64  | Import from Nepal (sparse) |
| remarks                 | object   | Special notes/events |
| temperature             | float64  | Temperature (°C) |
| humidity                | float64  | Humidity (%) |
| rain                    | float64  | Rainfall (mm) |
| windspeed               | float64  | Wind speed |

# Data Preprocessing

In [116]:
# Rename Columns for Consistency
df = df.rename(columns={
    'generation_mw': 'gen_total_mw',
    'load_shedding': 'load_shedding_mw',
    'gas': 'gen_gas_mw',
    'liquid_fuel': 'gen_liquid_fuel_mw',
    'coal': 'gen_coal_mw',
    'hydro': 'gen_hydro_mw',
    'solar': 'gen_solar_mw',
    'wind': 'gen_wind_mw',
    'india_bheramara_hvdc': 'import_india_bheramara_mw',
    'india_tripura': 'import_india_tripura_mw',
    'india_adani': 'import_india_adani_mw',
    'nepal': 'import_nepal_mw',
    'temperature': 'temp_c',
    'humidity': 'humidity_pct',
    'rain': 'rain_mm',
    'windspeed': 'wind_speed_ms'
})

In [117]:
df.head()

,datetime,gen_total_mw,demand_mw,load_shedding_mw,gen_gas_mw,gen_liquid_fuel_mw,gen_coal_mw,gen_hydro_mw,gen_solar_mw,gen_wind_mw,import_india_bheramara_mw,import_india_tripura_mw,import_india_adani_mw,import_nepal_mw,remarks,temp_c,humidity_pct,rain_mm,wind_speed_ms
0,2020-01-01 00:00:00,5913.0,5913,0,5493,0,160,47,0.0,NaN,129,76,NaN,NaN,NaN,18.3,74.0,0.0,3.5
1,2020-01-01 01:00:00,5570.0,5570,0,5117,0,160,47,0.0,NaN,129,74,NaN,NaN,NaN,16.5,81.0,0.0,6.0
2,2020-01-01 02:00:00,5351.0,5351,0,4900,0,159,47,0.0,NaN,129,72,NaN,NaN,NaN,14.6,89.0,0.0,8.4
3,2020-01-01 03:00:00,5223.0,5223,0,4773,0,160,46,0.0,NaN,129,70,NaN,NaN,NaN,14.5,90.0,0.0,6.3
4,2020-01-01 04:00:00,5142.0,5142,0,4697,0,159,46,0.0,NaN,129,68,NaN,NaN,NaN,14.1,90.0,0.0,8.2


In [118]:
# check null value 
df['datetime'].isna().sum()

np.int64(0)

In [119]:
# Convert datetime column
df['datetime'] = pd.to_datetime(df['datetime'])

In [120]:
# Sort values 
df = df.sort_values('datetime')

In [121]:
# Set index
df = df.set_index('datetime')

In [122]:
df.head()

,gen_total_mw,demand_mw,load_shedding_mw,gen_gas_mw,gen_liquid_fuel_mw,gen_coal_mw,gen_hydro_mw,gen_solar_mw,gen_wind_mw,import_india_bheramara_mw,import_india_tripura_mw,import_india_adani_mw,import_nepal_mw,remarks,temp_c,humidity_pct,rain_mm,wind_speed_ms
datetime,,,,,,,,,,,,,,,,,,
2020-01-01 00:00:00,5913.0,5913,0,5493,0,160,47,0.0,NaN,129,76,NaN,NaN,NaN,18.3,74.0,0.0,3.5
2020-01-01 01:00:00,5570.0,5570,0,5117,0,160,47,0.0,NaN,129,74,NaN,NaN,NaN,16.5,81.0,0.0,6.0
2020-01-01 02:00:00,5351.0,5351,0,4900,0,159,47,0.0,NaN,129,72,NaN,NaN,NaN,14.6,89.0,0.0,8.4
2020-01-01 03:00:00,5223.0,5223,0,4773,0,160,46,0.0,NaN,129,70,NaN,NaN,NaN,14.5,90.0,0.0,6.3
2020-01-01 04:00:00,5142.0,5142,0,4697,0,159,46,0.0,NaN,129,68,NaN,NaN,NaN,14.1,90.0,0.0,8.2


In [123]:
# check missing values
df.isnull().sum()

gen_total_mw                     0
demand_mw                        0
load_shedding_mw                 0
gen_gas_mw                       0
gen_liquid_fuel_mw               0
gen_coal_mw                      0
gen_hydro_mw                     0
gen_solar_mw                     0
gen_wind_mw                  31188
import_india_bheramara_mw        0
import_india_tripura_mw          0
import_india_adani_mw        42526
import_nepal_mw              44513
remarks                      46058
temp_c                        1991
humidity_pct                  1991
rain_mm                       1991
wind_speed_ms                 1991
dtype: int64

In [124]:
# missing values percentage
np.round(df.isna().mean()*100, 2)

gen_total_mw                  0.00
demand_mw                     0.00
load_shedding_mw              0.00
gen_gas_mw                    0.00
gen_liquid_fuel_mw            0.00
gen_coal_mw                   0.00
gen_hydro_mw                  0.00
gen_solar_mw                  0.00
gen_wind_mw                  62.56
import_india_bheramara_mw     0.00
import_india_tripura_mw       0.00
import_india_adani_mw        85.30
import_nepal_mw              89.29
remarks                      92.39
temp_c                        3.99
humidity_pct                  3.99
rain_mm                       3.99
wind_speed_ms                 3.99
dtype: float64

## Null Values in the Dataset

After checking the dataset, I found that **most columns have 0% null values**.

However, the following columns contain **constant non-zero values**, which might indicate placeholder or default data instead of actual collected values:
- `gen_wind_mw` - 62.56% of values are identical
- `import_india_adani_mw` - 85.30% of values are identical
- `import_nepal_mw` - 89.29% of values are identical
- `remarks` – 92.39% of values are identical
- `temp_c` – 3.99% of values are identical
- `humidity_pct` – 3.99% of values are identical
- `rain_mm` – 3.99% of values are identical
- `wind_speed_ms` – 3.99% of values are identical

In [125]:
# Check time range
print("Start:", df.index.min())
print("End:", df.index.max())

Start: 2020-01-01 00:00:00
End: 2025-06-17 00:00:00


## Handling Missing values

### **`gen_wind_mw`,`import_india_adani_mw`, and `import_nepal_mw` columns**
- The missing values in `gen_wind_mw`, `import_india_adani_mw`, and `import_nepal_mw` represent structural zeros rather than true missing data. These sources are inactive for large periods, so missing entries indicate no generation or import. Therefore, filling with zero preserves the real-world meaning of the data and avoids introducing artificial patterns into the model.

In [126]:
# fill missing values
df['gen_wind_mw'].fillna(0, inplace=True)
df['import_india_adani_mw'].fillna(0, inplace=True)
df['import_nepal_mw'].fillna(0, inplace=True)

C:\Users\dclid\AppData\Local\Temp\ipykernel_14480\811341114.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['gen_wind_mw'].fillna(0, inplace=True)
C:\Users\dclid\AppData\Local\Temp\ipykernel_14480\811341114.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, whe

### **`remarks` Column**

In [144]:
# find unique remaks value
df['remarks'].unique()

array([nan, 'Day_Peak', 'Evening_Peak'], dtype=object)

In [145]:
# create a new columns of hours
df['hour'] = df.index.hour

In [146]:
# find unique 'hours'
df['hour'].unique()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 23, 22], dtype=int32)

In [147]:
# check electicity demand for different hours
df.groupby(df.index.hour)['demand_mw'].mean()

datetime
0     10272.610025
1      9900.017060
2      9660.555165
3      9488.279980
4      9187.461423
5      8941.368104
6      8973.097438
7      9262.379569
8      9471.846308
9      9645.950401
10     9867.537381
11    10232.788867
12    10482.272682
13    10321.533133
14    10334.871178
15    10128.062720
16     9992.128822
17    10394.373747
18    11581.148808
19    11631.910866
20    11593.029087
21    11334.526817
22    11036.527081
23    10724.221665
Name: demand_mw, dtype: float64

### Decision Making
---
**Evening Peak (clearly highest)**
- Hour 18 → 11581
- Hour 19 → 11631 (peak)
- Hour 20 → 11593

18-20 is Evening_Peak

---
**Day Peak (moderate-high)**
- Hour 9–17 range is ~10,000–10,400

9–17 is Day_Peak

---
**Night Peak & Late_Evening**
- Hour 0–6 lowest values (~8900–9900)
0–6 is Night_Peak & 21–23 is Late_Evening
---

In [148]:
df['remarks'] = np.select(
    [
        df['hour'].between(18, 20),
        df['hour'].between(9, 17),
        df['hour'].between(0, 6),
        df['hour'].between(21, 23)
    ],
    [
        'Evening_Peak',
        'Day_Peak',
        'Night_OffPeak',
        'Late_Evening'
    ],
    default='Day_Peak'
)

In [152]:
df.isna().sum()

gen_total_mw                    0
demand_mw                       0
load_shedding_mw                0
gen_gas_mw                      0
gen_liquid_fuel_mw              0
gen_coal_mw                     0
gen_hydro_mw                    0
gen_solar_mw                    0
gen_wind_mw                     0
import_india_bheramara_mw       0
import_india_tripura_mw         0
import_india_adani_mw           0
import_nepal_mw                 0
remarks                         0
temp_c                       1991
humidity_pct                 1991
rain_mm                      1991
wind_speed_ms                1991
hour                            0
dtype: int64